// node-red-contrib-lf-nodes/nodes/addAlarm/addAlarm.js
module.exports = function(RED) {
  var fetch = require('node-fetch')

  function addAlarm(config) {
    RED.nodes.createNode(this, config)
    var context = this.context().global
    this.ID = config.ID
    this.URL = config.URL
    this.Device = config.Device
    this.Branch = config.Branch
    this.Object = config.Object
    this.Sensor = config.Sensor
    var node = this
    this.on('input', function(msg) {
      // var token = context.get('token')
      var defaultURL = 'https://lf.aexol.com/graphql' // 'http://localhost:8080/graphql'
      const data = msg.payload
      const mutation = `mutation {
        addAlarm (alarm: {
          Device:  "${node.Device}",
          Branch: "${node.Branch}",
          Object: "${node.Device}",
          Sensor: "${node.Sensor}",
          Value: "${data['Value']}",
          Status: [${data['Status']}]
        }),
        {
          Status,
          ID
        }
      }`
      fetch(node.URL || defaultURL, {
        method: 'POST',
        mode: 'cors',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({
          query: mutation
        })
      })
      .then(resp => resp.json())
      .then(resp2 => {
        msg.payload = JSON.stringify(resp2)
        node.send(msg)
      })
      .catch(err => {
        node.log(err)
      })
    })
  }
  RED.nodes.registerType("addAlarm", addAlarm)
}
