// node-red-contrib-lf-nodes/nodes/updateAlarmStatus/updateAlarmStatus.js
module.exports = function(RED) {
  var fetch = require('node-fetch')

  function updateAlarmStatus(config) {
    RED.nodes.createNode(this, config)
    var context = this.context().global
    this.URL = config.URL
    this.Device = config.Device
    var node = this
    this.on('input', function(msg) {
      // var token = context.get('token')
      var defaultURL = 'https://lf.aexol.com/graphql' // 'http://localhost:8080/graphql'
      const data = msg.payload
      const mutation = `mutation {
        changeStatusByDevice (deviceStatus: {
          Device:  "${node.Device}",
          status: [${data}]
        })
      }`
      fetch(node.URL || defaultURL, {
        method: 'POST',
        mode: 'cors',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({
          query: mutation
        })
      })
      .then(resp => resp.json())
      .then(resp2 => {
        msg.payload = JSON.stringify(resp2)
        node.send(msg)
      })
      .catch(err => {
        node.log(err)
      })
    })
  }
  RED.nodes.registerType("updateAlarmStatus", updateAlarmStatus)
}
